File: src/main/java/phoenix/world/structures/CustomEndSpikeConfig.java
package phoenix.world.structures;


import com.google.common.collect.ImmutableMap;
import com.mojang.datafixers.Dynamic;
import com.mojang.datafixers.types.DynamicOps;
import mcp.MethodsReturnNonnullByDefault;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.gen.feature.IFeatureConfig;

import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;
import java.util.List;
import java.util.stream.IntStream;
import java.util.stream.Stream;

@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
public class CustomEndSpikeConfig implements IFeatureConfig
{
    private final boolean crystalInvulnerable;
    private final List<CustomEndSpike.EndSpike> spikes;
    @Nullable
    private final BlockPos crystalBeamTarget;

    public CustomEndSpikeConfig(boolean crystalInvulnerable, List<CustomEndSpike.EndSpike> spikes, @Nullable BlockPos crystalBeamTarget) {
        this.crystalInvulnerable = crystalInvulnerable;
        this.spikes = spikes;
        this.crystalBeamTarget = crystalBeamTarget;
    }

    public <T> Dynamic<T> serialize(DynamicOps<T> ops) {
        return new Dynamic<>(ops, ops.createMap(ImmutableMap.of(ops.createString("crystalInvulnerable"),
                ops.createBoolean(this.crystalInvulnerable),
                ops.createString("spikes"),
                ops.createList(this.spikes.stream().map((spike) -> spike.serialise(ops).getValue())),
                ops.createString("crystalBeamTarget"),
                this.crystalBeamTarget == null ? ops.createList(Stream.empty()) : ops.createList(IntStream.of(this.crystalBeamTarget.getX(),
                this.crystalBeamTarget.getY(),
                this.crystalBeamTarget.getZ()).mapToObj(ops::createInt)))));
    }

    public static <T> CustomEndSpikeConfig deserialize(Dynamic<T> dynamic) {
        List<CustomEndSpike.EndSpike> list = dynamic.get("spikes").asList(CustomEndSpike.EndSpike::deserialize);
        List<Integer> list1 = dynamic.get("crystalBeamTarget").asList((d) -> d.asInt(0));
        BlockPos blockpos;
        if (list1.size() == 3) {
            blockpos = new BlockPos(list1.get(0), list1.get(1), list1.get(2));
        } else {
            blockpos = null;
        }

        return new CustomEndSpikeConfig(dynamic.get("crystalInvulnerable").asBoolean(false), list, blockpos);
    }

    public boolean isCrystalInvulnerable() {
        return this.crystalInvulnerable;
    }

    public List<CustomEndSpike.EndSpike> getSpikes() {
        return this.spikes;
    }

    @Nullable
    public BlockPos getCrystalBeamTarget() {
        return this.crystalBeamTarget;
    }
}

File: src/main/kotlin/phoenix/world/structures/CustomEndSpikeConfig.kt
package phoenix.world.structures

import com.google.common.collect.ImmutableMap
import com.mojang.datafixers.Dynamic
import com.mojang.datafixers.types.DynamicOps
import mcp.MethodsReturnNonnullByDefault
import net.minecraft.util.math.BlockPos
import net.minecraft.world.gen.feature.IFeatureConfig
import java.util.stream.IntStream
import java.util.stream.Stream
import javax.annotation.ParametersAreNonnullByDefault


@ParametersAreNonnullByDefault
@MethodsReturnNonnullByDefault
class CustomEndSpikeConfig(
    val isCrystalInvulnerable: Boolean,
    val spikes: List<CustomEndSpike.EndSpike>,
    val crystalBeamTarget: BlockPos?
) :
    IFeatureConfig
{

    override fun <T> serialize(ops: DynamicOps<T>): Dynamic<T>
    {
        return Dynamic(
            ops, ops.createMap(
                ImmutableMap.of(
                    ops.createString("crystalInvulnerable"),
                    ops.createBoolean(isCrystalInvulnerable),
                    ops.createString("spikes"),
                    ops.createList(spikes.stream().map { spike: CustomEndSpike.EndSpike ->
                        spike.serialise(
                            ops
                        ).value
                    }),
                    ops.createString("crystalBeamTarget"),
                    if (crystalBeamTarget == null) ops.createList(Stream.empty()) else ops.createList(
                        IntStream.of(
                            crystalBeamTarget.x,
                            crystalBeamTarget.y,
                            crystalBeamTarget.z
                        ).mapToObj { value: Int -> ops.createInt(value) })
                )
            )
        )
    }

    companion object
    {
        fun <T> deserialize(dynamic: Dynamic<T>): CustomEndSpikeConfig
        {
            val list = dynamic["spikes"].asList(CustomEndSpike.EndSpike.Companion::deserialize)

            val list1 = dynamic["crystalBeamTarget"].asList { d: Dynamic<T> -> d.asInt(0) }

            val blockpos: BlockPos? = if (list1.size == 3) BlockPos(list1[0], list1[1], list1[2]) else null

            return CustomEndSpikeConfig(dynamic["crystalInvulnerable"].asBoolean(false), list, blockpos)
        }
    }
}
